index_api-v2.md

```diff
@@ -0,0 +1,191 @@
+
+Using these two pieces of information, aggregators can then search the index for nodes with profiles that match specific schemas they want to use to create maps and directories, and the index returns a list of URLs (`profile_url`s) for the matched nodes' profiles. Aggregators then download from each node it's profile data for use in their apps.
+
+The index also stores additional data to enable aggregators to locate nodes based on other information in addition to their associated schemas. This additional information includes, for example:
+
+- _Geolocation data_ - latitude/longitude of primary location of entity - enables searching for nodes within a geographical range (i.e., "1 kilometer from my current location" or "60 miles from my home")
+- _Location data_ - town/city, country, etc. for searching based on map location
+- _Tags_ - nodes can include an array of strings ('tags') that describe the node (e.g., 'restaurant', 'pizza', 'beer', 'calzones')
+<!-- TODO - add this back in if/when a solution is determined
+- _Entity type_ - Wikipedia/Wikidata URL that best describes the organization type - enables searching for specific types of entities (e.g., food co-ops: [https://en.wikipedia.org/wiki/Food_cooperative](https://en.wikipedia.org/wiki/Food_cooperative))
+-->
+
+> :construction: INDEX SYNCING
+>
+> A `profile_hash` is for future planning, and will also be stored for each node profile so that multiple indices can synchronize their lists of nodes. It will be a hash of the profile (the entire JSON object containing all node profile data) stored at the `profile_url` that was submitted by the node. Nodes can compare `node_id`s (the unique hash of a `profile_url`) and then `profile_hash`es between themselves. They will also store a Unix timestamp (`last_updated`) of when they last validated the profile at the `profile_url` and created the `profile_hash` for each `node_id`. The index with the oldest timestamp should download the profile from the `profile_url` and regenerate the `profile_hash`. This is just a rough idea of index syncing; it will need to be thought through in more detail, including thinking about edge cases.
+
+## Node Endpoints
+
+### [`POST /validate`](https://app.swaggerhub.com/apis-docs/MurmurationsNetwork/IndexAPI/2.0.0#/Node%20Endpoints/post_validate)
+
+Node operators can test the validity of a profile to one or more schemas before posting the profile to their website and submitting the `profile_url` to the index.
+
+#### Input
+
+- A valid JSON schema instance (the profile).
+
+#### Output
+
+- 200 OK response (if profile validates to the schema(s))
+- 400 Bad Request response (if profile does not validate) along with one or more failure reasons
```

Comment:
It may be too late to change this in the current round of updates, but I don't think we should use 400 responses in this case. This should return a 200 response, because the purpose of this endpoint is to check a profile that may or may not be valid. A request to this endpoint that tells the node that the profile is invalid is still a successful request (it achieved the purpose of the endpoint). A 400 error, according to MDN:

"indicates that the server cannot or will not process the request due to something that is perceived to be a client error (for example, malformed request syntax, invalid request message framing, or deceptive request routing)." 

In our case, for this endpoint an invalid profile is not an invalid request in any of the senses above. 

On the practical side, using 4xx responses can cause problems on the client, because systems that manage the HTTP(S) communications rightly treat these responses as indications of an error at the request level that should never occur, and treat them as exceptions that need to be handled using a different code flow than normal results. As in the case of the previous situation of 404s for empty search results, awkward coding is required to work around this.

Reply:
OK. I updated the docs to reflect this and created an issue:
https://github.com/MurmurationsNetwork/MurmurationsServices/issues/265

We'll make the change before we release v2 into production.